### ISSUE TITLE
with_dict incorrectly passing null values as empty strings

### ISSUE THREAD

**Issue by gbolo** [ansible/ansible#17159]
##### ISSUE TYPE
- Feature Idea
- Bug Report

##### COMPONENT NAME

`with_dict`
##### ANSIBLE VERSION

```
ansible --version
ansible 2.1.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

N/A
##### OS / ENVIRONMENT

Running Ansible from: Arch Linux
Target Host: Centos 7, Arch Linux
##### SUMMARY

When passing a dictionary via `with_dict` in a task and using certain portions of it to create values of another dictionary variable, nulls are not being preserved and incorrectly passed as empty string ""
##### STEPS TO REPRODUCE

```
- hosts: localhost
  gather_facts: no
  vars:
        # defaults (for single nodes)
        param1: "defaultParam1"
        param2: ~
        param3: ""
        param4: []

        # overrides (for multiple nodes)
        nodes:
          node1:
            config_path: /tmp/node1.conf
            config_overrides:
              param1: "node1Param1"
              param2: ~
              param3: ""
              param4: []
          node2:
            config_path: /tmp/node2.conf
            config_overrides:

        # dictionaries to be converted to json in config
        single_node_config:
          Param1: "{{ param1 }}"
          Param2: "{{ param2 }}"
          Param3: "{{ param3 }}"
          Param4: "{{ param4 }}"

        multi_node_config:
          Param1: "{{ item.value.config_overrides.param1 | default(param1) }}"
          Param2: "{{ item.value.config_overrides.param2 | default(param2) }}"
          Param3: "{{ item.value.config_overrides.param3 | default(param3) }}"
          Param4: "{{ item.value.config_overrides.param4 | default(param4) }}"

        # toggle
        enable_multi_nodes: false
  tasks:
        # task for single node
        - name: create config
          copy:
            content: "{{ single_node_config | to_nice_json }}"
            dest: /tmp/single.conf
          when: enable_multi_nodes != true

        # task multi nodes
        - name: create configs
          copy:
            content: "{{ multi_node_config | to_nice_json }}"
            dest: "{{ item.value.config_path }}"
          with_dict: "{{ nodes }}"
          when: enable_multi_nodes == true
```
##### EXPECTED RESULTS

```
cat /tmp/single.conf /tmp/node1.conf  /tmp/node2.conf
{
    "Param1": "defaultParam1",
    "Param2": null,
    "Param3": "",
    "Param4": []
}
{
    "Param1": "node1Param1",
    "Param2": null,
    "Param3": "",
    "Param4": []
}
{
    "Param1": "defaultParam1",
    "Param2": null,
    "Param3": "",
    "Param4": []
}
```
##### ACTUAL RESULTS

```
cat /tmp/single.conf /tmp/node1.conf  /tmp/node2.conf
{
    "Param1": "defaultParam1",
    "Param2": null,
    "Param3": "",
    "Param4": []
}
{
    "Param1": "node1Param1",
    "Param2": "",
    "Param3": "",
    "Param4": []
}
{
    "Param1": "defaultParam1",
    "Param2": "",
    "Param3": "",
    "Param4": []
}
```

**Comment by agaffney** [comment]
FYI, @sivel

**Comment by agaffney** [comment]
Here is a much simpler test case showing the same issue in 2.1.1.0:

```
- hosts: localhost
  gather_facts: no
  vars:
        param2: ~

        foo:
          bar: ~
          baz: "{{ param2 }}"
  tasks:
        - debug: msg="{{ param2 }}"
        - debug: msg="{{ foo }}"
        - debug: msg="{{ foo.bar }}"
```

Result:

```
TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": null
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": {
        "bar": null,
        "baz": null
    }
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": ""
}
```

**Comment by gbolo** [comment]
This also seems to happen when you specify a default; the second task in this example loses it's `null`:

```
- hosts: localhost
  gather_facts: no
  vars:
        param2: ~

        foo:
          bar: ~
          baz: "{{ param2 }}"
  tasks:
        - debug: msg="{{ param2 }}"
        - debug: msg="{{ param2 | default('test') }}"
        - debug: msg="{{ foo }}"
        - debug: msg="{{ foo.bar }}"
```

Result:

```
TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": null
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": ""
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": {
        "bar": null, 
        "baz": null
    }
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": ""
}
```

**Comment by gbolo** [comment]
This bug also seems to affect integers:

```
- hosts: localhost
  gather_facts: no
  vars:
        param2: 7

        foo:
          bar: "7"
          baz: 7
  tasks:
        - debug: msg="{{ param2 }}"
        - debug: msg="{{ param2 | default('test') }}"
        - debug: msg="{{ foo }}"
        - debug: msg="{{ foo.bar }}"
        - debug: msg="{{ foo.baz }}"
```

Result:

```
TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": 7
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "7"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": {
        "bar": "7", 
        "baz": 7
    }
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "7"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "7"
}
```

**Comment by alikins** [comment]
I can reproduce with ~2.2

```
ansible 2.2.0 (devel 3e754086b6) last updated 2016/09/16 15:14:47 (GMT -400)
  lib/ansible/modules/core: (detached HEAD 2e1e3562b9) last updated 2016/09/16 15:14:47 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 9b5c64e240) last updated 2016/09/16 15:14:47 (GMT -400)
  config file = /home/adrian/ansible/ansible.cfg
  configured module search path = ['/home/adrian/src/ansible-modules-core', '/home/adrian/src/ansible-modules-extras']
```

```
[fedora-23:ansible (devel % u=)]$ cat repro/with_dict_none_17159/integers.yml 
- hosts: localhost
  gather_facts: no
  vars:
        param2: 7

        foo:
          bar: "7"
          baz: 7
  tasks:
      - name: debug param2 should be int 7
        debug: msg="{{ param2 }}"
      - name: debug param2 or default
        debug: msg="{{ param2 | default('test') }}"
      - name: debug foo
        debug: msg="{{ foo }}"
      - name: debug foo.bar
        debug: msg="{{ foo.bar }}"
      - name: debug foo.baz should be int 7 not string
        debug: msg="{{ foo.baz }}"

[fedora-23:ansible (devel % u=)]$ ansible-playbook -vvv repro/with_dict_none_17159/integers.yml 
Using /home/adrian/ansible/ansible.cfg as config file
[DEPRECATION WARNING]: vault.VaultLib.is_encrypted is deprecated.  Use vault.is_encrypted instead.
This feature will be removed in version 2.4. Deprecation warnings can be disabled by setting 
deprecation_warnings=False in ansible.cfg.

PLAYBOOK: integers.yml *********************************************************
1 plays in repro/with_dict_none_17159/integers.yml

PLAY [localhost] ***************************************************************

TASK [debug param2 should be int 7] ********************************************
task path: /home/adrian/src/ansible/repro/with_dict_none_17159/integers.yml:10
ok: [localhost] => {
    "msg": 7
}

TASK [debug param2 or default] *************************************************
task path: /home/adrian/src/ansible/repro/with_dict_none_17159/integers.yml:12
ok: [localhost] => {
    "msg": "7"
}

TASK [debug foo] ***************************************************************
task path: /home/adrian/src/ansible/repro/with_dict_none_17159/integers.yml:14
ok: [localhost] => {
    "msg": {
        "bar": "7", 
        "baz": 7
    }
}

TASK [debug foo.bar] ***********************************************************
task path: /home/adrian/src/ansible/repro/with_dict_none_17159/integers.yml:16
ok: [localhost] => {
    "msg": "7"
}

TASK [debug foo.baz should be int 7 not string] ********************************
task path: /home/adrian/src/ansible/repro/with_dict_none_17159/integers.yml:18
ok: [localhost] => {
    "msg": "7"
}

PLAY RECAP *********************************************************************
localhost                  : ok=5    changed=0    unreachable=0    failed=0   
```

```
[fedora-23:ansible (devel % u=)]$ cat repro/with_dict_none_17159/simple.yml
- hosts: localhost
  gather_facts: no
  vars:
        param2: ~

        foo:
          bar: ~
          baz: "{{ param2 }}"
  tasks:
        - debug: msg="{{ param2 }}"
        - debug: msg="{{ foo }}"
        - name: debug foo.bar should be None
          debug: msg="{{ foo.bar }}"
        - name: debug foo.baz should be None
          debug: msg="{{ foo.baz }}"

[fedora-23:ansible (devel % u=)]$ ansible-playbook -vvv repro/with_dict_none_17159/simple.yml 
Using /home/adrian/ansible/ansible.cfg as config file
[DEPRECATION WARNING]: vault.VaultLib.is_encrypted is deprecated.  Use vault.is_encrypted instead.
This feature will be removed in version 2.4. Deprecation warnings can be disabled by setting 
deprecation_warnings=False in ansible.cfg.

PLAYBOOK: simple.yml ***********************************************************
1 plays in repro/with_dict_none_17159/simple.yml

PLAY [localhost] ***************************************************************

TASK [debug] *******************************************************************
task path: /home/adrian/src/ansible/repro/with_dict_none_17159/simple.yml:10
ok: [localhost] => {
    "msg": null
}

TASK [debug] *******************************************************************
task path: /home/adrian/src/ansible/repro/with_dict_none_17159/simple.yml:11
ok: [localhost] => {
    "msg": {
        "bar": null, 
        "baz": null
    }
}

TASK [debug foo.bar should be None] ********************************************
task path: /home/adrian/src/ansible/repro/with_dict_none_17159/simple.yml:12
ok: [localhost] => {
    "msg": ""
}

TASK [debug foo.baz should be None] ********************************************
task path: /home/adrian/src/ansible/repro/with_dict_none_17159/simple.yml:14
ok: [localhost] => {
    "msg": ""
}

PLAY RECAP *********************************************************************
localhost                  : ok=4    changed=0    unreachable=0    failed=0   

```

**Comment by gbolo** [comment]
would be nice to get this fixed for 2.2 release

**Comment by alikins** [comment]
More oddity. I'm starting to wonder if this is a bug in 'debug'...

debug seems to show the values as strings, but assert thinks they are none

[fedora-23:ansible (devel % u=)]$ cat repro/with_dict_none_17159/integers.yml 

``` yaml

- hosts: localhost
  gather_facts: no
  vars:
        param2_6_int: 6

        foo:
          bar_7_string: "7"
          baz_8_int: 8
  tasks:
      - name: debug param2 should be int 6
        debug: msg="{{ param2_6_int }}"

      - name: debug param2 or default
        debug: msg="{{ param2_6_int | default('test') }}"

      - name: debug foo
        debug: msg="{{ foo }}"

      - name: debug foo.bar_7_string
        debug: msg="{{ foo.bar_7_string }}"

      - name: assert foo.baz_8_int is an int
        assert:
            that:
                - foo.baz_8_int == 8
                - foo.bar_7_string == "7"
                - foo.bar_7_string != 7
                - foo.baz_8_int != "8"

      - name: debug foo.baz_8_int_i should be int 8 not string
        debug: msg="{{ foo.baz_8_int }}"
```

```
fedora-23:ansible (devel % u=)]$ ansible-playbook repro/with_dict_none_17159/integers.yml 

PLAY [localhost] ***************************************************************

TASK [debug param2 should be int 6] ********************************************
ok: [localhost] => {
    "msg": 6
}

TASK [debug param2 or default] *************************************************
ok: [localhost] => {
    "msg": "6"
}

TASK [debug foo] ***************************************************************
ok: [localhost] => {
    "msg": {
        "bar_7_string": "7", 
        "baz_8_int": 8
    }
}

TASK [debug foo.bar_7_string] **************************************************
ok: [localhost] => {
    "msg": "7"
}

TASK [assert foo.baz_8_int is an int] ******************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "All assertions passed"
}

TASK [debug foo.baz_8_int_i should be int 8 not string] ************************
ok: [localhost] => {
    "msg": "8"
}

PLAY RECAP *********************************************************************
localhost                  : ok=6    changed=0    unreachable=0    failed=0   
```

[fedora-23:ansible (devel % u=)]$ cat repro/with_dict_none_17159/test.yml 

``` yaml
- hosts: localhost
  gather_facts: no
  vars:
        param2: ~

        foo:
          bar: ~
          baz: "{{ param2 }}"
  tasks:
        - assert:
            that:
                - param2 == none
                - param2 is none
                - param2 is defined
                - param2 != ""

        - debug: msg="{{ param2 }}"
        - debug: msg="{{ foo }}"

# foo.bar should be null, but ends up as empty string
        - debug: msg="foo.bar should be |null| but is |{{ foo.bar }}|"

        - assert:
            that:
                - foo.bar == none
                - foo.bar is none
                - foo.bar is defined
                - foo.bar != ""

        - debug: msg="foo.baz should be |null| but is |{{ foo.baz }}|"

        - assert:
            that:
                - foo.baz == none
                - foo.baz is none
                - foo.baz is defined
                - foo.baz != ""
```

```
[fedora-23:ansible (devel % u=)]$ ansible-playbook repro/with_dict_none_17159/test.yml 

PLAY [localhost] ***************************************************************

TASK [assert] ******************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "All assertions passed"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": null
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": {
        "bar": null, 
        "baz": null
    }
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "foo.bar should be |null| but is ||"
}

TASK [assert] ******************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "All assertions passed"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "foo.baz should be |null| but is ||"
}

TASK [assert] ******************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "All assertions passed"
}

PLAY RECAP *********************************************************************
localhost                  : ok=7    changed=0    unreachable=0    failed=0   

```

**Comment by jctanner** [comment]
@gbolo what are you expecting to happen in this case? ...

```
- debug: msg="{{ param2 |  default('test') }}"
```

needs_info

**Comment by alikins** [comment]
I think the ~ being shown as empty string is intentional, and has existed since at least 2.0.

@jctanner found the commit that added a finalizer to never return None/null from jinja2
https://github.com/ansible/ansible/blame/devel/lib/ansible/template/__init__.py#L413-L417

**Comment by alikins** [comment]
Testing with:

``` yaml
- hosts: localhost
  gather_facts: no
  vars:
        param2: ~

        foo:
          bar: ~
          baz: "{{ param2 }}"
  tasks:
        - debug: msg="{{ param2 }}"
        - debug: msg="{{ foo }}"
        - name: debug foo.bar should be None
          debug: msg="{{ foo.bar }}"
        - name: debug foo.baz should be None
          debug: msg="{{ foo.baz }}"
```

Shows the same behavior for 2.2,2.1, and 2.0.

```
[fedora-23:ansible-2.0.2.0-1 (v2.0.2.0-1-branch)]$ ansible-playbook ~/src/ansible/repro/with_dict_none_17159/simple.yml 

PLAY [localhost] ***************************************************************

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": null
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": {
        "bar": null, 
        "baz": null
    }
}

TASK [debug foo.bar should be None] ********************************************
ok: [localhost] => {
    "msg": ""
}

TASK [debug foo.baz should be None] ********************************************
ok: [localhost] => {
    "msg": ""
}

PLAY RECAP *********************************************************************
localhost                  : ok=4    changed=0    unreachable=0    failed=0   
```

```
[fedora-23:ansible-stable-2.1 (stable-2.1 u=)]$ ansible-playbook ~/src/ansible/repro/with_dict_none_17159/simple.yml 

PLAY [localhost] ***************************************************************

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": null
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": {
        "bar": null, 
        "baz": null
    }
}

TASK [debug foo.bar should be None] ********************************************
ok: [localhost] => {
    "msg": ""
}

TASK [debug foo.baz should be None] ********************************************
ok: [localhost] => {
    "msg": ""
}

PLAY RECAP *********************************************************************
localhost                  : ok=4    changed=0    unreachable=0    failed=0   
```

```
[fedora-23:ansible (devel % u=)]$ ansible --version
ansible 2.2.0 (devel dde7c11713) last updated 2016/09/22 10:25:35 (GMT -400)
  lib/ansible/modules/core: (detached HEAD 0f505378c3) last updated 2016/09/22 10:25:36 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 935a3ab2cb) last updated 2016/09/22 10:25:36 (GMT -400)
  config file = /home/adrian/ansible/ansible.cfg
  configured module search path = Default w/o overrides
[fedora-23:ansible (devel % u=)]$ ansible-playbook ~/src/ansible/repro/with_dict_none_17159/simple.yml 



PLAY [localhost] ***************************************************************

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": null
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": {
        "bar": null, 
        "baz": null
    }
}

TASK [debug foo.bar should be None] ********************************************
ok: [localhost] => {
    "msg": ""
}

TASK [debug foo.baz should be None] ********************************************
ok: [localhost] => {
    "msg": ""
}

PLAY RECAP *********************************************************************
localhost                  : ok=4    changed=0    unreachable=0    failed=0   
```

**Comment by gbolo** [comment]
@jctanner 

```
- debug: msg="{{ param2 |  default('test') }}"
```

I expect to not lose the null. An empty string is different from null. The problem I have having, is that i use dictionaries to create json configs with the jinja2 filter `to_nice_json`. Currently, these issues make this filter useless for providing accurate json config files for applications which need accurate json files. 

In summary, both `null` and `int` types seem to be affected, since ansible is passing strings to represent both.

**Comment by gbolo** [comment]
@alikins are you saying that this will not be fixed? what about the integers changing to strings?

**Comment by jctanner** [comment]
A python snippet and example output to clarify some of the behaviors described in this issue ...
https://gist.github.com/jctanner/7d1faf7109606c08dfc4608aeea42c3f

If we have two variables such as ...

```
param2 = None
param3 = 'None'
```

And then create jinja template expressions to render those variables ...

```
{{ param2 }}
{{ param3 }}
```

We would get the following result from a vanilla jinja configuration...

```
u'None'
u'None'
```

In ansible's jinja configuraiton, we would get ...

```
u'None'
u''
```

The last example is the crux of this issue. When a null variable not inside a datastructure is templated by ansible, it becomes an empty string. The reason is that jinja would otherwise return u'None', so ansible makes use of the "finalize" parameter during instantiation of the jinja environment ...

https://github.com/ansible/ansible/blob/1f311f07395a721119dc22a46f684f19a999309a/lib/ansible/template/__init__.py#L166

```
        self.environment = Environment(
            trim_blocks=True,
            undefined=StrictUndefined,
            extensions=self._get_extensions(),
            finalize=self._finalize,
            loader=FileSystemLoader(self._basedir),
        )
```

The _finalize function is ...

```
return thing if thing is not None else ''
```

The finalize function has been in place since November of 2014, so the behavior is -not- a regression. Presumably the code was written because we can't know if jinja was given a null or a 'None', so we tell jinja to truncate None to empty string rather than getting back u'None' for null and None.

There's no bug here in my opinion, but there is a complication in how templates can be used to represent null values. We will have to do further research and possibly figure out how to overcome that.

**Comment by gbolo** [comment]
@jctanner 

Thanks for spending some time on this issue. I look forward to the possibility of solution to this shortcoming. 

There still remains the issue with integers being converted to strings as described above. Perhaps a separate issue should be open for it?

**Comment by alikins** [comment]
> @alikins are you saying that this will not be fixed? 

Not sure it's broken since it seems to be intentional behavior. And identity and equality checks
internal to playbooks seem to work.

> what about the integers changing to strings?

Still seems busted to me, but I suspect the cause is different.

> - debug: msg="{{ param2 |  default('test') }}"

In that case, I think param2 is defined (to null), so the default doesn't get used (based on http://jinja.pocoo.org/docs/dev/templates/#default). But that does point to the option of

```
debug: msg="{{ param2|default('test', true) }}"
```

which will fall back to 'test' since param2 is falsey (defined, but null). That is limited
to vars that are not expected to be falsey though.

Easiest approach may just to not define 'param2' for a case like this (or 'param3' for this ex):

```
        - name: debug param3 not defined
          debug: msg="{{ param3|default('the_param3_default') }}"
```

There is also a 'none' jinja test, that is a little verbose to test:

```
        - name: debug none(param2)
          debug: msg="{% if param2 is none %}test{% else %} {{ param }} {% endif %}"
```

**Comment by ansibot** [comment]
@gbolo ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@gbolo ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jctanner** [comment]
!needs_info

**Comment by samdoran** [comment]
This is expected behavior as described [here](https://github.com/ansible/ansible/issues/57438#issuecomment-499559239). Use [native Jinja types](https://docs.ansible.com/ansible/latest/reference_appendices/config.html#default-jinja2-native) to change this behavior.

Duplicate of #41392 and #57438
